File: src/web/llm/gateway.ts
/**
 * gateway.ts — unified LLM client wrapping the LiteLLM gateway.
 *
 * All chat completions in the app funnel through here. The gateway is an
 * OpenAI-compatible endpoint (LiteLLM) that routes to Azure OpenAI or any
 * configured fallback (see infra/litellm/config.yaml).
 *
 * Design goals:
 *   - Model name is a gateway alias ("chat-primary"), never a provider
 *     deployment name. Physical routing is a server-side concern.
 *   - graceful degradation: on error / missing API key, return null and let
 *     the caller decide the fallback UX (matches existing agent behaviour).
 *   - lightweight retry: 3 attempts w/ exponential backoff on network / 5xx.
 */

import OpenAI from "openai";
import { config } from "../../config";

// ---------- Types ----------

export type ChatMessage = {
  role: "system" | "user" | "assistant";
  content: string;
};

export type ChatOptions = {
  system: string;
  user: string;
  /** Gateway model alias. Defaults to config.llm.chatModel ("chat-primary"). */
  model?: string;
  temperature?: number;
  maxTokens?: number;
  /** AbortController.signal for cancellation. */
  signal?: AbortSignal;
};

// ---------- Client (lazy singleton) ----------

let _client: OpenAI | null = null;

function getClient(): OpenAI | null {
  if (!config.llm.apiKey || !config.llm.baseUrl) {
    return null;
  }
  if (_client) return _client;
  _client = new OpenAI({
    baseURL: config.llm.baseUrl.replace(/\/$/, "") + "/v1",
    apiKey: config.llm.apiKey,
    // OpenAI SDK does its own retry; we still wrap with a manual retry loop
    // below for finer-grained control.
    maxRetries: 0,
  });
  return _client;
}

/**
 * True if the gateway has enough config to attempt a call. Agents use this
 * to decide whether to fall back to their non-LLM code paths, mirroring the
 * pre-migration behaviour that gated on AZURE_OPENAI_ENDPOINT / API_KEY.
 */
export function isGatewayConfigured(): boolean {
  return Boolean(config.llm.apiKey && config.llm.baseUrl);
}

// ---------- Retry helper ----------

const RETRY_STATUSES = new Set([408, 429, 500, 502, 503, 504]);

async function withRetry<T>(fn: () => Promise<T>): Promise<T> {
  const delays = [500, 1000, 2000];
  let lastErr: unknown;
  for (let attempt = 0; attempt < delays.length + 1; attempt++) {
    try {
      return await fn();
    } catch (err) {
      lastErr = err;
      // Retry only on retryable HTTP statuses / network errors
      const status = (err as { status?: number })?.status;
      const isRetryable =
        status === undefined /* network */ ||
        RETRY_STATUSES.has(status);
      if (!isRetryable || attempt === delays.length) {
        throw err;
      }
      // Respect Retry-After if present (OpenAI SDK exposes it via headers)
      const headers = (err as { headers?: Record<string, string> })?.headers;
      const retryAfter = headers?.["retry-after"]
        ? Number(headers["retry-after"]) * 1000
        : delays[attempt];
      await new Promise((r) => setTimeout(r, retryAfter));
    }
  }
  throw lastErr;
}

// ---------- Public API ----------

/**
 * chatText — plain-text chat completion. Returns the assistant's text or null
 * on any failure (matches the pre-migration `callAzureOpenAI` contract).
 */
export async function chatText(opts: ChatOptions): Promise<string | null> {
  const client = getClient();
  if (!client) return null;

  const model = opts.model ?? config.llm.chatModel;

  try {
    return await withRetry(async () => {
      const res = await client.chat.completions.create(
        {
          model,
          messages: [
            { role: "system", content: opts.system },
            { role: "user", content: opts.user },
          ],
          temperature: opts.temperature ?? 0.3,
          max_completion_tokens: opts.maxTokens ?? 800,
        },
        { signal: opts.signal }
      );
      return res.choices?.[0]?.message?.content?.trim() ?? null;
    });
  } catch (err) {
    console.error("[llm/gateway] chatText failed:", err);
    return null;
  }
}

/**
 * chatJson — chat completion with `response_format: json_object`.
 * Returns the raw JSON string (parsing is caller's responsibility, so
 * validation logic stays close to each domain type).
 */
export async function chatJson(opts: ChatOptions): Promise<string | null> {
  const client = getClient();
  if (!client) return null;

  const model = opts.model ?? config.llm.chatModel;

  try {
    return await withRetry(async () => {
      const res = await client.chat.completions.create(
        {
          model,
          messages: [
            { role: "system", content: opts.system },
            { role: "user", content: opts.user },
          ],
          temperature: opts.temperature ?? 0.2,
          max_completion_tokens: opts.maxTokens ?? 1500,
          response_format: { type: "json_object" },
        },
        { signal: opts.signal }
      );
      return res.choices?.[0]?.message?.content?.trim() ?? null;
    });
  } catch (err) {
    console.error("[llm/gateway] chatJson failed:", err);
    return null;
  }
}

File: server/app/llm/gateway.py
"""Chat completion client that talks to the LiteLLM gateway.

Direct port of ``src/web/llm/gateway.ts``. Contract:

* ``chat_text`` / ``chat_json`` return the assistant text on success or
  ``None`` on any failure (missing config, network error, non-retryable HTTP).
* Retries are bounded to 3 attempts on 408 / 429 / 5xx / network errors, with
  exponential backoff (500ms, 1s, 2s) and ``Retry-After`` honoured.
"""

from __future__ import annotations

import asyncio
import logging
from typing import Any, Literal, Optional

import httpx
from openai import APIError, APIStatusError, AsyncOpenAI

from app.config import settings

logger = logging.getLogger("app.llm.gateway")

RETRY_STATUSES = {408, 429, 500, 502, 503, 504}
_BACKOFF_MS = (500, 1000, 2000)

Role = Literal["system", "user", "assistant"]

_client: Optional[AsyncOpenAI] = None


def is_gateway_configured() -> bool:
    """True iff we have enough env to attempt a call."""
    return bool(settings.llm.api_key and settings.llm.base_url)


def _get_client() -> Optional[AsyncOpenAI]:
    global _client
    if not is_gateway_configured():
        return None
    if _client is not None:
        return _client
    base = settings.llm.base_url.rstrip("/") + "/v1"
    _client = AsyncOpenAI(
        base_url=base,
        api_key=settings.llm.api_key,
        max_retries=0,  # we handle retries manually below
        timeout=httpx.Timeout(60.0, connect=10.0),
    )
    return _client


async def _with_retry(fn):
    """Retry ``fn()`` on retryable errors."""
    last_exc: Optional[BaseException] = None
    for attempt in range(len(_BACKOFF_MS) + 1):
        try:
            return await fn()
        except APIStatusError as exc:
            last_exc = exc
            status = exc.status_code
            if status not in RETRY_STATUSES or attempt == len(_BACKOFF_MS):
                raise
            retry_after_hdr = exc.response.headers.get("retry-after") if exc.response else None
            delay = (
                float(retry_after_hdr) * 1000
                if retry_after_hdr and retry_after_hdr.isdigit()
                else _BACKOFF_MS[attempt]
            )
            await asyncio.sleep(delay / 1000)
        except (APIError, httpx.RequestError, asyncio.TimeoutError) as exc:
            last_exc = exc
            if attempt == len(_BACKOFF_MS):
                raise
            await asyncio.sleep(_BACKOFF_MS[attempt] / 1000)
    assert last_exc is not None
    raise last_exc


async def chat_text(
    *,
    system: str,
    user: str,
    model: Optional[str] = None,
    temperature: float = 0.3,
    max_tokens: int = 800,
) -> Optional[str]:
    """Plain-text chat completion. Returns text or ``None`` on failure."""
    client = _get_client()
    if client is None:
        return None

    model_name = model or settings.llm.chat_model

    async def _call() -> Optional[str]:
        res = await client.chat.completions.create(
            model=model_name,
            messages=[
                {"role": "system", "content": system},
                {"role": "user", "content": user},
            ],
            temperature=temperature,
            max_completion_tokens=max_tokens,
        )
        content = res.choices[0].message.content if res.choices else None
        return content.strip() if content else None

    try:
        return await _with_retry(_call)
    except Exception as exc:  # noqa: BLE001
        logger.error("chat_text failed: %s", exc)
        return None


async def chat_json(
    *,
    system: str,
    user: str,
    model: Optional[str] = None,
    temperature: float = 0.2,
    max_tokens: int = 1500,
) -> Optional[str]:
    """JSON-mode chat completion. Returns the raw JSON string or ``None``."""
    client = _get_client()
    if client is None:
        return None

    model_name = model or settings.llm.chat_model

    async def _call() -> Optional[str]:
        res = await client.chat.completions.create(
            model=model_name,
            messages=[
                {"role": "system", "content": system},
                {"role": "user", "content": user},
            ],
            temperature=temperature,
            max_completion_tokens=max_tokens,
            response_format={"type": "json_object"},
        )
        content = res.choices[0].message.content if res.choices else None
        return content.strip() if content else None

    try:
        return await _with_retry(_call)
    except Exception as exc:  # noqa: BLE001
        logger.error("chat_json failed: %s", exc)
        return None


__all__ = [
    "chat_json",
    "chat_text",
    "is_gateway_configured",
]

# Silence pyright unused-import in Any/Role helpers if they're introduced later.
_ = Any
_ = Role
